applications/RomApplication/custom_utilities/rom_residuals_utility.h

```diff
@@ -0,0 +1,189 @@
+                            if (dofs[i*mNodalDofs + k]->IsFixed())
+                                row(PhiElemental, i*mNodalDofs + k) = ZeroVector(PhiElemental.size2());
+                            else
+                                row(PhiElemental, i*mNodalDofs+k) = row(rom_nodal_basis,k);
+                        }
+                    }
+                    ResidualReduced = prod(trans(PhiElemental), RHS_Contribution);
+                    row(MatrixResiduals, k) = ResidualReduced;
+                    
+                    // clean local elemental me overridemory
+                    mpScheme->CleanMemory(*(it_el.base()));
+                }
+
+            }
+
+            for (int k = 0; k < nconditions;  k++){
+                ModelPart::ConditionsContainerType::iterator it = cond_begin + k;
+                //detect if the condition is active or not. If the user did not make any choice the condition
+                //is active by default
+                bool condition_is_active = true;
+                if ((it)->IsDefined(ACTIVE))
+                    condition_is_active = (it)->Is(ACTIVE);
+                if (condition_is_active){
+                    Condition::DofsVectorType dofs;
+                    it->GetDofList(dofs, CurrentProcessInfo);
+                    //calculate elemental contribution
+                    mpScheme->Condition_CalculateSystemContributions(*(it.base()), LHS_Contribution, RHS_Contribution, EquationId, CurrentProcessInfo);
+                    //assemble the elemental contribution - here is where the ROM acts
+                    //compute the elemental reduction matrix PhiElemental
+                    const auto& r_geom = it->GetGeometry();
+                    Matrix PhiElemental(r_geom.size()*mNodalDofs, mRomDofs);
+                    Vector ResidualReduced(mRomDofs); // The size of the residual will vary only when using more ROM modes, one row per condition
+                    for(unsigned int i=0; i<r_geom.size(); ++i){
+                        const Matrix& rom_nodal_basis = r_geom[i].GetValue(ROM_BASIS);
+                        for(unsigned int k=0; k<rom_nodal_basis.size1(); ++k){
+                            if (dofs[i*mNodalDofs + k]->IsFixed())
+                                row(PhiElemental, i*mNodalDofs + k) = ZeroVector(PhiElemental.size2());
+                            else
+                                row(PhiElemental, i*mNodalDofs+k) = row(rom_nodal_basis,k);
+                        }
+                    }
+                    ResidualReduced = prod(trans(PhiElemental), RHS_Contribution);
+                    row(MatrixResiduals, k+nelements) = ResidualReduced;
+
+                    // clean local elemental memory
+                    mpScheme->CleanMemory(*(it.base()));
+                }
+            }
+        return MatrixResiduals;
+        }
+
+        void convert_to_numpy(const Matrix & KratosMatrix, pybind11::object NumpyMatrix)
```

Comment:
this function looks quite complicated and manual to me.
Have you checked out the pybind interface with numpy? => https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html
Even easier would probably be to take a look at how @oberbichler did the Ublas-Eigen conversion in the EigenSolversApp and then directly use the Eigen-numpy wrappers that pybind provides https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html

Reply:
BTW cannot you use the scipy-conversion things @RiccardoRossi added recently?
Those would probably do the same, since you anyway copy the values

Reply:
I did try using Riccardo's implementation on #6096 . However, this matrix is dense, and it did not work. 
About the other implementations you mention, I will be checking them..